### ISSUE TITLE
Bug in cache copy assignment operator

### ISSUE THREAD

**Issue by unixod** [uni-algo/uni-algo#37]
Hi,

```
uni-algo/include/uni_algo/internal/ranges_core.h: In instantiation of ‘constexpr una::detail::rng::cache<T>& una::detail::rng::cache<T>::operator=(const una::detail::rng::cache<T>&) [with T = una::ranges::grapheme::utf8_view<std::__cxx11::basic_string<char> >::utf8<__gnu_cxx::__normal_iterator<char*, std::__cxx11::basic_string<char> >, __gnu_cxx::__normal_iterator<char*, std::__cxx11::basic_string<char> > >]’:
uni-algo/include/uni_algo/ranges_grapheme.h:27:7:   required from here
uni-algo/include/uni_algo/internal/ranges_core.h:166:17: error: ‘class una::ranges::grapheme::utf8_view<std::__cxx11::basic_string<char> >::utf8<__gnu_cxx::__normal_iterator<char*, std::__cxx11::basic_string<char> >, __gnu_cxx::__normal_iterator<char*, std::__cxx11::basic_string<char> > >’ has no member named ‘has_val’
  166 |             val.has_val = false;
      |             ~~~~^~~~~~~
```

**Comment by mg152** [comment]
I cannot reproduce it. What compiler are you using?

**Comment by unixod** [comment]
> I cannot reproduce it. What compiler are you using?

gcc 13.2.0

Take a look at cache copy assignment operator, there is one obvious and one not-quite obvious bugs in it:

```c++
constexpr cache& operator=(const cache& other) noexcept
{
    if (std::addressof(other) != this)
        val.has_val = false;    <<< 1. should be other.has_val
                                <<< 2. absence of  has_val = false for this;
    return *this;
}
```

**Comment by mg152** [comment]
Ah, I see now. There is a typo that breaks compilations only in rare cases when this operator= is actually used. Must be `has_val = false` there, and `other.has_val = false` is not needed according to C++ Standard.

Should be fixed in commit a9f9d77400cb7ce24ebfb7088a3fd20b4458eb9d. Thanks for reporting this!

**Comment by unixod** [comment]
Thanks!
